quantecon/markov/approximation.py

```diff
@@ -212,3 +214,200 @@ def _fill_tauchen(x, P, n, rho, sigma, half_step):
+        An array where element (i,j) of S is the discretized
+        value of the j-th element of x_t in state i. Reducing S to its
+        unique values yields the grid values.
+    Xvec : array_like(float)
+        A matrix of size m x sim_length containing the simulated time
+        series of the m discretized states.
+
+
+    Notes
+    -----
+        The code presently assumes normal shocks but normality is not required
+        for the algorithm to work. The draws from the multivariate standard
+        normal generator can be replaced by any other random number generator
+        with mean 0 and unit standard deviation.
+
+
+    Example
+    -------
+
+        This example discretizes the stochastic process used to calibrate
+        the economic model included in ``Downward Nominal Wage Rigidity,
+        Currency Pegs, and Involuntary Unemployment'' by Stephanie
+        Schmitt-Grohé and Martín Uribe, Journal of Political Economy 124,
+        October 2016, 1466-1514.
+
+            A     = np.array([[0.7901, -1.3570],
+                              [-0.0104, 0.8638]])
+            Omega = np.array([[0.0012346, -0.0000776],
+                              [-0.0000776, 0.0000401]])
+            grid_sizes = np.array([21, 11])
+            Pi, Xvec, S = discrete_var(A, Omega, grid_sizes,
+                              sim_length=1_000_000, burn_in = 100_000)
+    """
+
+    m = len(A)   # The number of dimensions of the original state x_t
+    default_grid_size = 10
+
+    if grid_sizes is None:
+        # Set the size of every grid to default_grid_size
+        grid_sizes = np.full(m, default_grid_size)
+
+    n = grid_sizes.prod()  # Size of the discretized state
+
+    # Compute stationary variance-covariance matrix of AR process and use
+    # it to obtain grid bounds.
+    Sigma = qme.solve_discrete_lyapunov(A, Omega)
+    sigma_vector = np.sqrt(np.diagonal(Sigma))    # Stationary std dev
+    upper_bounds = std_devs * sigma_vector
+
+    # Build the individual grids along each dimension
+    V = []
+    for i in range(m):
+        b = np.linspace(-upper_bounds[i], upper_bounds[i], grid_sizes[i])
+        V.append(b)
+
+    S = cartesian_product(V, row_major=row_major)
+
+    Pi = np.zeros((n, n))
+    Xvec = np.zeros((m, sim_length))
+    C = sp.linalg.sqrtm(Omega)
+
+    # Run simulation to compute transition probabilities
+    _run_sim(A, C, Pi, Xvec, S, sim_length, burn_in, seed)
+
+    # Cut states where the column sum of Pi is zero (i.e., inaccesible states
+    # according to the simulation)
+    indx = np.where(np.sum(Pi, axis=0) > 0)
+    Pi = Pi[indx[0], :]
+    Pi = Pi[:, indx[0]]
+    S  = S[indx[0], :]
+
+    # Normalize
+    sum_row = np.sum(Pi, axis=1)
+    for i in range(len(Pi)):
+        Pi[i, :] = Pi[i, :] / sum_row[i]
+
+    if return_sim:
+        return Pi, S, Xvec
+    return Pi, S
+
+
+@njit
+def _run_sim(A, C, Pi, Xvec, S, sim_length, burn_in, seed):
+    m = len(A)
+    np.random.seed(seed)
```

Comment:
I do not know what the whole code does, but I do not think this is a good strategy.

Maybe better to follow the pattern used throughout the package (see e.g. [this](https://github.com/QuantEcon/QuantEcon.py/blob/89a70127b1321444b8f22ba262c7401f4b08e9c4/quantecon/markov/core.py#L471)) to have the outer function `discrete_var` accept a `random_state`, and
* generate random values within `discrete_var` by `random_state.randn(m, sim_length)` (or something appropriate),
* store them in `Xvec` (after multiplying by `C`?), and
* modify `Xvec` in place in `_run_sim`?

Reply:
Sorry, I haven't read https://github.com/QuantEcon/QuantEcon.py/pull/640#issuecomment-1295750336. Yes agreed, let an existing object do a simulation.

Reply:
Sorry, I haven't read https://github.com/QuantEcon/QuantEcon.py/pull/640#issuecomment-1295750336. Yes agreed, let an existing object do a simulation.